src/cluster.c

```diff
@@ -1,1491 +1,517 @@
+                     * keys/channels when the slot is in importing state. */
+                    multiple_keys = 1;
+                }
+            }
 
-    /* Turn into master. */
-    if (nodeIsSlave(myself)) {
-        clusterSetNodeAsMaster(myself);
-        replicationUnsetMaster();
-        emptyData(-1,EMPTYDB_NO_FLAGS,NULL);
+            /* Migrating / Importing slot? Count keys we don't have.
+             * If it is pubsubshard command, it isn't required to check
+             * the channel being present or not in the node during the
+             * slot migration, the channel will be served from the source
+             * node until the migration completes with CLUSTER SETSLOT <slot>
+             * NODE <node-id>. */
+            int flags = LOOKUP_NOTOUCH | LOOKUP_NOSTATS | LOOKUP_NONOTIFY | LOOKUP_NOEXPIRE;
+            if ((migrating_slot || importing_slot) && !is_pubsubshard)
+            {
+                if (lookupKeyReadWithFlags(&server.db[0], thiskey, flags) == NULL) missing_keys++;
+                else existing_keys++;
+            }
+        }
+        getKeysFreeResult(&result);
     }
 
-    /* Close slots, reset manual failover state. */
-    clusterCloseAllSlots();
-    resetManualFailover();
-
-    /* Unassign all the slots. */
-    for (j = 0; j < CLUSTER_SLOTS; j++) clusterDelSlot(j);
-
-    /* Recreate shards dict */
-    dictEmpty(server.cluster->shards, NULL);
-
-    /* Forget all the nodes, but myself. */
-    di = dictGetSafeIterator(server.cluster->nodes);
-    while((de = dictNext(di)) != NULL) {
-        clusterNode *node = dictGetVal(de);
-
-        if (node == myself) continue;
-        clusterDelNode(node);
-    }
-    dictReleaseIterator(di);
+    /* No key at all in command? then we can serve the request
+     * without redirections or errors in all the cases. */
+    if (n == NULL) return myself;
 
-    /* Hard reset only: set epochs to 0, change node ID. */
-    if (hard) {
-        sds oldname;
-
-        server.cluster->currentEpoch = 0;
-        server.cluster->lastVoteEpoch = 0;
-        myself->configEpoch = 0;
-        serverLog(LL_NOTICE, "configEpoch set to 0 via CLUSTER RESET HARD");
-
-        /* To change the Node ID we need to remove the old name from the
-         * nodes table, change the ID, and re-add back with new name. */
-        oldname = sdsnewlen(myself->name, CLUSTER_NAMELEN);
-        dictDelete(server.cluster->nodes,oldname);
-        sdsfree(oldname);
-        getRandomHexChars(myself->name, CLUSTER_NAMELEN);
-        getRandomHexChars(myself->shard_id, CLUSTER_NAMELEN);
-        clusterAddNode(myself);
-        serverLog(LL_NOTICE,"Node hard reset, now I'm %.40s", myself->name);
+    uint64_t cmd_flags = getCommandFlags(c);
+    /* Cluster is globally down but we got keys? We only serve the request
+     * if it is a read command and when allow_reads_when_down is enabled. */
+    if (!isClusterHealthy()) {
```

Comment:
I would posit that I think there is utility in this having 3 states:
1. I can serve all traffic. (AFAIK everything is healthy in the cluster so I can serve slots based on my known state)
2. I can serve reads. (I'm not sure about the cluster state, but I know I still own my slots)
3. I should serve no traffic. (I have no idea what is going on, maybe the cluster is completing migrated and I should serve nothing)

I think for example "cluster_allow_reads_when_down" is really a statement about 2, but flotilla might have a different perspective about when a node should be able to serve reads.

Reply:
This seems reasonable. I mainly worked on a 1-1 translation/reorganization of the concepts into an interface that would allow us two separate (legacy and flotilla) imlementations, not on improvements. Of couse I am open to improvements but would prefer, not die on the hill of, keeping this PR focused as a reorganization, rather than making logical changes to the code.